clients/venice-consumer/src/main/java/com/linkedin/venice/client/consumer/VeniceAfterImageConsumerImpl.java

```diff
@@ -127,9 +164,26 @@ private PubSubMessage<K, ChangeEvent<V>, Long> convertRecordToPubSubMessage(
       case PUT:
         Put put = (Put) kafkaMessageEnvelope.payloadUnion;
         Schema valueSchema = schemaReader.getValueSchema(put.schemaId);
-        RecordDeserializer<V> valueDeserializer =
-            FastSerializerDeserializerFactory.getFastAvroGenericDeserializer(valueSchema, valueSchema);
-        currentValue = valueDeserializer.deserialize(put.putValue);
+        try {
+          currentValue = (V) bufferAndAssembleRecordChangeEvent(
+              pubSubTopicPartition,
+              put.getSchemaId(),
+              currentKey,
+              currentCompressor.decompress(put.getPutValue()).array(),
+              offset,
+              CHUNKING_ADAPTER,
+              Lazy.of(() -> FastSerializerDeserializerFactory.getFastAvroGenericDeserializer(valueSchema, valueSchema)),
+              put.schemaId,
+              readOnlySchemaRepository);
+          if (currentValue == null) {
+            // This was an event which we had to buffer, so don't return a result. We'll terminate early here as we
+            // should
+            // only do book keeping on consumed offset watermarks for those events which we return to the client.
+            return null;
+          }
+        } catch (IOException e) {
+          throw new RuntimeException(e);
```

Comment:
nit: Maybe VeniceException as we use across all other implementations, although RuntimeException is also fine to me.

Reply:
Seems fine, replacing.